File: src/EfsAiHub.Core.Abstractions/Observability/ProjectAnalytics.cs
namespace EfsAiHub.Core.Abstractions.Observability;

/// <summary>
/// Resumo agregado de uso/custo do projeto num intervalo. Alimenta o card
/// principal do dashboard (overview MTD).
/// </summary>
public sealed class ProjectOverview
{
    public required string ProjectId { get; init; }
    public required DateTime PeriodFrom { get; init; }
    public required DateTime PeriodTo { get; init; }
    public required decimal TotalCostUsd { get; init; }
    public required long TotalTokens { get; init; }
    public required int TotalCalls { get; init; }
    public required int TotalExecutions { get; init; }
    public required int Completed { get; init; }
    public required int Failed { get; init; }
    /// <summary>Sucesso/(Sucesso+Falha). Quando TotalExecutions==0 retorna 0.</summary>
    public required double SuccessRate { get; init; }
    public required IReadOnlyList<AgentMiniRow> TopAgents { get; init; }
}

/// <summary>Linha compacta usada em listas top-N (overview.topAgents).</summary>
public sealed class AgentMiniRow
{
    public required string AgentId { get; init; }
    public required long TotalTokens { get; init; }
    public required decimal CostUsd { get; init; }
    public required int Calls { get; init; }
}

/// <summary>Bucket temporal do gráfico de série (custo + execuções).</summary>
public sealed class ProjectTimeseriesBucket
{
    public required DateTime Bucket { get; init; }
    public required decimal CostUsd { get; init; }
    public required long Tokens { get; init; }
    public required int Calls { get; init; }
    public required int Executions { get; init; }
    public required int Completed { get; init; }
    public required int Failed { get; init; }
}

/// <summary>Linha da tabela de breakdown por agente do projeto.</summary>
public sealed class ProjectAgentBreakdown
{
    public required string AgentId { get; init; }
    public string? AgentName { get; init; }
    public string? ModelId { get; init; }
    public required int Calls { get; init; }
    public required long TotalTokens { get; init; }
    public required decimal CostUsd { get; init; }
    public required double AvgDurationMs { get; init; }
    public required double P95DurationMs { get; init; }
    /// <summary>Falhas/(Sucessos+Falhas). 0 quando sem execuções.</summary>
    public required double ErrorRate { get; init; }
}

/// <summary>
/// Estado do orçamento diário do projeto. Lê limites de ProjectSettings e
/// consumo do dia em Redis (já populado pelo ProjectBudgetGuard).
/// Hoje é warning-only no runtime — a UI deve sinalizar `exceeded:true` mas
/// avisar que a plataforma não bloqueia execução.
/// </summary>
public sealed class ProjectBudgetStatus
{
    public required string ProjectId { get; init; }
    public int? MaxTokensPerDay { get; init; }
    public decimal? MaxCostUsdPerDay { get; init; }
    public required long TodayTokens { get; init; }
    public required decimal TodayCostUsd { get; init; }
    /// <summary>0..1 (ou >1 quando estourou). Null quando o limite não está configurado.</summary>
    public double? TokensUsagePct { get; init; }
    public double? CostUsagePct { get; init; }
    /// <summary>true quando qualquer um dos limites configurados foi cruzado.</summary>
    public required bool Exceeded { get; init; }
}

File: mvp/public/lib/project-analytics.js
// @ts-check
/**
 * API helpers pra analytics do projeto. Substitui
 * mvp/src/api/projectAnalytics.ts.
 */

import { get } from './api.js';

/**
 * @typedef {object} AgentMiniRow
 * @property {string} agentId
 * @property {number} totalTokens
 * @property {number} costUsd
 * @property {number} calls
 *
 * @typedef {object} ProjectOverview
 * @property {string} projectId
 * @property {string} periodFrom
 * @property {string} periodTo
 * @property {number} totalCostUsd
 * @property {number} totalTokens
 * @property {number} totalCalls
 * @property {number} totalExecutions
 * @property {number} completed
 * @property {number} failed
 * @property {number} successRate
 * @property {AgentMiniRow[]} topAgents
 *
 * @typedef {object} ProjectTimeseriesBucket
 * @property {string} bucket
 * @property {number} costUsd
 * @property {number} tokens
 * @property {number} calls
 * @property {number} executions
 * @property {number} completed
 * @property {number} failed
 *
 * @typedef {object} ProjectAgentBreakdown
 * @property {string} agentId
 * @property {string | null} [agentName]
 * @property {string | null} [modelId]
 * @property {number} calls
 * @property {number} totalTokens
 * @property {number} costUsd
 * @property {number} avgDurationMs
 * @property {number} p95DurationMs
 * @property {number} errorRate
 *
 * @typedef {object} ProjectBudgetStatus
 * @property {string} projectId
 * @property {number | null} [maxTokensPerDay]
 * @property {number | null} [maxCostUsdPerDay]
 * @property {number} todayTokens
 * @property {number} todayCostUsd
 * @property {number | null} [tokensUsagePct]
 * @property {number | null} [costUsagePct]
 * @property {boolean} exceeded
 */

/**
 * @param {string} projectId
 * @returns {Promise<ProjectOverview>}
 */
export function getProjectOverview(projectId) {
  return get(`/analytics/projects/${encodeURIComponent(projectId)}/overview`);
}

/**
 * @param {string} projectId
 * @param {'day' | 'hour'} [groupBy]
 * @param {readonly string[]} [excludeAgentIds]
 * @returns {Promise<ProjectTimeseriesBucket[]>}
 */
export function getProjectTimeseries(projectId, groupBy = 'day', excludeAgentIds) {
  const qs = new URLSearchParams({ groupBy });
  if (excludeAgentIds && excludeAgentIds.length > 0) {
    qs.set('excludeAgentIds', excludeAgentIds.join(','));
  }
  return get(`/analytics/projects/${encodeURIComponent(projectId)}/timeseries?${qs}`);
}

/**
 * @param {string} projectId
 * @param {number} [top]
 * @returns {Promise<ProjectAgentBreakdown[]>}
 */
export function getProjectAgents(projectId, top = 20) {
  return get(`/analytics/projects/${encodeURIComponent(projectId)}/agents?top=${top}`);
}

/**
 * @param {string} projectId
 * @returns {Promise<ProjectBudgetStatus>}
 */
export function getProjectBudget(projectId) {
  return get(`/analytics/projects/${encodeURIComponent(projectId)}/budget`);
}
